cdap-app-templates/cdap-etl/cdap-data-pipeline/src/main/java/co/cask/cdap/datapipeline/LineageOperationsValidator.java

```diff
@@ -0,0 +1,275 @@
+
+      // We cannot simply do "unusedOutputs.removeAll(stageInputOutput.getOutputs()))"
+      // Consider following case assuming d is part of output schema:
+      // OP1: [a, b] -> [d]
+      // OP2: [b] -> [d]
+      // Here outout d from OP1 is redundant, since the d in output schema will always come from OP2.
+      // However d will not be in the redundantOutputs map, as we only put the redundant fields if they
+      // appear in input of some operation. Such redundancy should cause validation checks to fail.
+      Iterator<Map.Entry<String, List<String>>> iterator = unusedOutputs.entrySet().iterator();
+      while (iterator.hasNext()) {
+        Map.Entry<String, List<String>> next = iterator.next();
+        String field = next.getKey();
+        List<String> origins = next.getValue();
+
+        if (stageInputOutput.getOutputs().contains(field)) {
+          iterator.remove();
+          if (origins.size() > 1) {
+            // size will be greate than 1 only if its redundant
+            unusedOutputs.put(field, origins.subList(0, origins.size() - 1));
+          }
+        }
+      }
+
+      this.invalidOutputs.putAll(unusedOutputs);
+      this.invalidOutputs.putAll(redundantOutputs);
+    }
+
+    public Map<String, List<String>> getInvalidOutputs() {
+      return invalidOutputs;
+    }
+
+    public Map<String, List<String>> getInvalidInputs() {
+      return invalidInputs;
+    }
+
+    /**
+     * Validate the input fields.
+     *
+     * @param operationName name of the current operation
+     * @param inputsToValidate input fields for the current operation to be validated
+     * @param validInputsSoFar cumulative set of valid input fields, this contains the
+     *                         fields belonging to the input schema and outputs of all
+     *                         operations happening before the current operation
+     */
+    private void validateInputs(String operationName, List<String> inputsToValidate, Set<String> validInputsSoFar) {
+      for (String field : inputsToValidate) {
+        // check if field is valid input. input is valid if it is in the validInputsSoFar set
+        if (validInputsSoFar.contains(field)) {
+          continue;
+        }
+        List<String> originsWithInvalidInput = invalidInputs.computeIfAbsent(field, k -> new ArrayList<>());
+        originsWithInvalidInput.add(operationName);
+      }
+    }
+
+    /**
+     * Remove valid outputs from the map outputs which are yet to validate.
+     *
+     * @param operationInputs inputs to the operation, these can be marked as used and can be removed
+     *                        from outputsToValidate
+     * @param outputsToValidate cumulative outputs which are not yet validated
+     * @param redundantOutputs redundant outputs will be added to this map
+     */
+    private void validateOutputs(List<String> operationInputs, Map<String, List<String>> outputsToValidate,
```

Comment:
it's more accurate to call this updateInvalidOutputs.

Reply:
done.